### Handin 1


# Info

Everything should be completed and approved in person. Groups are fine.

The objectives for this handin is:
* Playing with numpy.
* Plotting with Plotly (as a preparation for using dash (https://plotly.com/dash/) later on)
* Train a more complex model using the Fortuna Algorithm.
* Route planning for Knut Knut Transport AS. 

## How to solve?
There should be a comment/section stating the objective of a task.  
And there should be a commented section labeled # -- CODE --  
that shows where to put your code. For tasks that are not answered by code you can either write the answer  
as a markup cell or write it on a seperate piece of paper.

In [ ]:
import numpy as np


# TASK: Create a numpy array containing the whole numbers between 1, 10 (inclusive)

# -- CODE --

a = np.arange(1, 11)


assert isinstance(a, np.ndarray)
assert np.isclose(a, [1,2,3,4,5,6,7,8,9,10]).all()
print("<ok>")

<ok>


In [ ]:
# TASK: Reshape the array a so that it is a 2 by 4 array

a = np.array(range(0, 8))

# -- CODE --
a = a.reshape(2, 4)

assert( a.shape == (2,4) )
print("<ok>")

<ok>


In [ ]:
# TASK: multiply all the numbers in a by 2.


a = np.array([[1,2,3,4], [5,6,7,8]])

# -- CODE --
a = a * 2

assert a.shape == (2,4)
assert a.sum() == 72
print("<ok>")

<ok>


In [ ]:
# TASK:  create a numpy array b that contains the sum of each row (axis 1) in a 

a = np.array([[1,2,3,4], [5,6,7,8]])

# -- CODE --
b = a.sum(axis=1)

assert b.shape == (2,)
assert (b == [10, 26]).all()
print("<ok>")

<ok>


In [ ]:
# TASK:  create a numpy array b that contains the mean value of each column (axis 0) in a 
a = np.array([[1,2,3,4], [1,2,3,8]])

# -- CODE --
b = a.mean(axis=0)

assert b.shape == (4,)
assert (b == [1, 2, 3, 6]).all()
print("<ok>")

<ok>


In [ ]:
# TASK:  stack the arrays a, b and c into an array called s (vertically)
# s should be equal to s_true (hint, look at the numpy function vstack )
import numpy as np

a = np.array([1, 3])
b = np.array([2, 4])
c = np.array([3, 5])

# -- CODE --
s = np.vstack((a, b, c))

s_true = np.array([[1, 3], [2, 4], [3, 5]])
assert (s_true == s).all()

print("<ok>")

<ok>


In [ ]:
# TASK:  Flatten the array a into a 1d array called f (hint: numpy has a function named flatten)

a = np.random.randn(2,2,2)
print(a)

# -- CODE --
f = a.flatten()

print("-"*50)
print(f)
assert f.ndim == 1
print("<ok>")

[[[-1.23315713  0.30968122]
  [ 0.71363709 -1.56303183]]

 [[ 0.11325489  1.61468583]
  [-0.79149906  0.4020182 ]]]
--------------------------------------------------
[-1.23315713  0.30968122  0.71363709 -1.56303183  0.11325489  1.61468583
 -0.79149906  0.4020182 ]
<ok>


## Task 1 -- Plot functions and scatter plots using Plotly 

Change the code to perform the two following tasks:

(See Plotly: https://plotly.com/python/plotly-express/ for more info. If you for some reason cannot get plotly to work
just use another plotting framework).

**Line Plot)**

Plot the function f(x) = -0.69x^2 + 1.3x + 0.42 over the interval [0, 2.5], with 0.01 increments in x.

**Scatter Plot)**  

Plot the first 25 Fibbonachi numbers using a scatter plot.
(The example shows the first 5).




In [ ]:
# installs plotly
!pip install plotly
!pip install jupyterlab ipywidgets

print("You may have to restart jupyter to see graphs in the notebook.")

You may have to restart jupyter to see graphs in the notebook.


In [4]:
import numpy as np
import plotly.express as px


# Change this code into the correct Line Plot
xs = np.arange(0,2.5, 0.01)

ys = -0.69 * xs**2 + 1.3 * xs + 0.42

fig = px.line(x=xs, y=ys, title="Line Plot")
fig.update_layout(xaxis_range=[xs.min(),xs.max()], yaxis_range=[ys.min(),ys.max()])
fig.show()

In [7]:
# Change this code into the correct Scatter Plot (please take care of the axis)
# Plot the first 25 Fibbonachi numbers using a scatter plot.
# (The example shows the first 5).
# xs = np.array([1, 2, 3, 4, 5])
# ys = np.array([1, 1, 2, 3, 5])

# xs = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25])
# ys = np.array([1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025])

n = 25
# crée un tableau vide de taille n pour stocker les nombres de Fibonacci (entier et pas flotant)
fib = np.empty(n, dtype=np.int64)
fib[0] = fib[1] = 1
for i in range(2, n):
    fib[i] = fib[i-1] + fib[i-2]

xs = np.arange(1, n + 1)
ys = fib


fig = px.scatter(x=xs, y=ys, title="Fibbonachi Scatter Plot")
fig.update_layout(xaxis_range=[xs.min(),xs.max()], yaxis_range=[ys.min(),ys.max()])
fig.show()

# Task 2 -- Numpy and curve fitting

Fill in the function get_2polynomal_from_3_points(). You are required to use the  
function: np.linalg.solve() for the actual computation, i.e. the task is mainly   
about shaping the input so it fits the np.linalg.solve() function. Note: dont use polyfit().

The function should return a 1d numpy array, with shape = (3,) in the format a, b,c.  
Such that ax^2 + bx + c = y, goes through the points p0, p1, p2

HINT: How should a matrix equation be formed, so solving it gives us the coefficients for a polynomial?

Limitations:
* There should be NO loops or ifs! (for/while/if). 
* Numpy functions that might come handy: reshape, hstack, flatten, ones



In [2]:
import numpy as np

def get_2polynomal_from_3_points(p0: np.array, p1: np.array, p2: np.array) -> np.array:
    # -- CODE --
    points = np.vstack((p0, p1, p2))
    print ("points:", points)
    x = points[:, 0]
    y = points[:, 1]
    # print("x:", x)
    # print("y:", y)
   
    col_x2 = (x**2).reshape(3, 1)   # [[x0²], [x1²], [x2²]]
    col_x  = x.reshape(3, 1)        # [[x0],  [x1],  [x2]]
    col_1  = np.ones((3, 1))        # [[1],   [1],   [1]]

    A = np.hstack((col_x2, col_x, col_1))  # [[x0², x0, 1], [x1², x1, 1], [x2², x2, 1]]
    print("A:", A)

    coefficients = np.linalg.solve(A, y)  # Solve for [a, b, c] in Ax = y
    return coefficients

    
point0 = np.array([0.147, 0.596])
point1 = np.array([0.7, 0.992])
point2 = np.array([2.06, 0.17])

polynomial_coefficients = get_2polynomal_from_3_points(point0, point1, point2)
assert(polynomial_coefficients.shape[0] == 3 and polynomial_coefficients.ndim == 1)

print("Polynomial: {:2f}x^2 + {:.2f}x + {:.2f}".format(*polynomial_coefficients))


if not np.isclose(polynomial_coefficients[0], -0.69, atol=0.01):
    print("x^2 coeff is wrong (a), should be close to -0.69")
    
if not np.isclose(polynomial_coefficients[1], 1.3, atol=0.01):
    print("x coeff is wrong (b), should be close to 1.3")
    
if not np.isclose(polynomial_coefficients[2], 0.42, atol=0.01):
    print("Constant factor (c) is wrong, should be close to 0.42")
    
if np.isclose(polynomial_coefficients, [-0.69, 1.3, 0.42], atol=0.01).all():
    print("Correct polynomial found.")

points: [[0.147 0.596]
 [0.7   0.992]
 [2.06  0.17 ]]
A: [[0.021609 0.147    1.      ]
 [0.49     0.7      1.      ]
 [4.2436   2.06     1.      ]]
Polynomial: -0.690280x^2 + 1.30x + 0.42
Correct polynomial found.


# Task 3 -- Simple random search

Find the triplet $a,b,c \in \{x \;|\; x \in \mathbb{Z} \text{ and } 450 > x > 0 \}$ 

Using a random search in the parameter space. Such that the following relations is satisfied: 

### a
$a = \begin{cases} c+11, & \text{if } b\text{ is even} \\ 2c-129, & \text{if } b\text{ is odd} \end{cases}$

### b
$b = (a \times c) \mod 2377$

### c
$c = \left( \sum\limits_{k=0}^{a-1} b - 7k \right) + 142$

**Also how many guesses were needed?**

Note that in math notation $\sum\limits_{k=1}^{5}k = 1+2+3+4+5$ 

Hint (to check if you got the right answer): The product $abc$ should be 255450 $\pm$ 20 
    
Can you find any optimizations besides multi-thread/multi-process?
Note: This HAS to be solved by a random search, not pen and paper.

In [3]:
# TASK: random search for a, b, c in {1, ..., 449} satisfying the three relations.
# Also report how many guesses were needed.

import random


def isValid(a, b, c):
    if a % 2 == 0:
        a_expected = c + 11
    else: 
        a_expected = 2*c - 129;

    if a != a_expected:
        return False

    # équation 2
    if b != (a * c) % 2377:
        return False

    # équation 3
    if c != sum(b - 7*k for k in range(a)) + 142:
        return False

    return True

tentatives = 0
while True:
    a = random.randint(1, 449)
    b = random.randint(1, 449)
    c = random.randint(1, 449)
    tentatives += 1
    if isValid(a, b, c):
        break

print("a, b, c =", a, b, c)
print("a * b * c =", a * b * c)
print("tentatives =", tentatives)

a, b, c = 31 103 80
a * b * c = 255440
tentatives = 34571775


# Task 4 -- Revisit the Fortuna Algorithm

Below is a implementation of Fortuna that uses fits linear function $f_{\theta} = a x + b$ where $\theta = \{a,b\}$ to a function $g(x)$.  
However, as is evidently from the graph, $g(x)$ is not a linear function.  

1) Change the code to instead use $f_{\theta}(x) = \sum\limits_{k=1}^{3} \Psi_k \sin(\gamma_k (x + \omega_k)) $.  
  Such that $|\theta| = 9$, where each parameter $c$ in $\theta$ is in $[-4. 4]$.  
That is, $f_{\theta}(x)$ is a sum of three sin terms. **Do not change the range of the sample\_theta function.**

3) However, it seems Fortuna (on average) struggles to find the optimal parameters $\theta$.  
Therefore you will have to innovate and change the Fortuna algorithm so that it faster finds "better solutions".
What changes did you make and **why** did you make them, and how did you measure how efficient these changes were?  
A excellent solution here will have an expected best loss of less than 5 using 100000 guesses. (take the average over 100 runs).
**But ANY improvment is sufficient to pass!**  

4) Using your newly made modified Fortuna Algorithm optimize the function: $h(x) = \mu - (\zeta sin(\kappa x) )  (\tau (x + \lambda))$ .  
The y values for this function can be found in the numpy array ys_h_1 (in the code below).  
Also test the values for ys_h_2. (the x is the same for all functions).

Does your new and improved Fortuna outperform the regular fortuna on this function as well? Why?   
**Remember to change your model to match $h(x)$**


4) [**Optional**] Develop a multiprocces implemention of the Fortuna algorithm using python's multiprocessing library (https://docs.python.org/3/library/multiprocessing.html).  
How are the speed ups? Are Fortuna really suited to parallel execution? (note: dont try multiprocess in a notebook - use a standalone file)




In [4]:
%%time
# the magic %%time has to be the first line in the cell, and reports the total exec time of the cell.

import random
import numpy as np
import plotly.express as px
import tqdm


def predict(x, theta):
    # theta contains 9 values: 3 parameters (psi, gamma, omega) for each sine
    psi1, gamma1, omega1, psi2, gamma2, omega2, psi3, gamma3, omega3 = theta

    #  psi1 * np.sin(gamma1 * (x + omega1)) =>  ps1 = amplitude ; np.sin = base fn ; gamma1 = frequency ; x + omega1 = phase shift
    return (psi1 * np.sin(gamma1 * (x + omega1))
          + psi2 * np.sin(gamma2 * (x + omega2))
          + psi3 * np.sin(gamma3 * (x + omega3)))


def sample_theta(size_of_theta):
    # Do NOT CHANGE.
    theta = np.random.uniform(-4, 4, size=size_of_theta)
    return theta


def get_loss(y_hat, ys):
    # No change needed, returns quadratic loss.
    loss = ((y_hat - ys)**2).sum()
    return loss

xs = np.array([1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.3, 7.4])
ys = np.array([4.03, 4.19, 4.26, 4.25, 4.17, 4.03, 3.85, 3.63, 3.40, 3.16, 2.93, 2.72, 2.53, 2.39, 2.28, 2.21, 2.18, 2.19, 2.22, 2.27, 2.33, 2.39, 2.44, 2.45, 2.43, 2.36, 2.22, 2.02, 1.75, 1.41, 1.00, 0.52, -0.01, -0.60, -1.22, -1.86, -2.50, -3.13, -3.72, -4.27, -4.75, -5.15, -5.45, -5.65, -5.74, -5.70, -5.55, -5.29, -4.92, -4.44, -3.89, -3.26, -2.58, -1.86, -1.12, -0.39, 0.32, 0.98, 1.60, 2.14, 2.61, 2.99, 3.28, 3.47, 3.57])

# 
ys_h_1 = np.array([15.98, 21.42, 24.1, 23.87, 21.0, 16.11, 10.06, 3.79, -1.8, -6.01, -8.39, -8.82, -7.47, -4.77, -1.3, 2.31, 5.49, 7.81, 9.04, 9.18, 8.44, 7.15, 5.72, 4.54, 3.89, 3.9, 4.52, 5.54, 6.63, 7.39, 7.49, 6.7, 4.97, 2.44, -0.54, -3.48, -5.8, -6.98, -6.61, -4.51, -0.79, 4.2, 9.8, 15.21, 19.57, 22.09, 22.2, 19.63, 14.52, 7.41, -0.83, -9.07, -16.11, -20.83, -22.36, -20.27, -14.59, -5.91, 4.72, 15.9, 26.06, 33.69, 37.58, 36.94, 31.64])
ys_h_2 = np.array([[5.87, 5.83, 5.74, 5.62, 5.48, 16.58, 18.21, 18.49, 17.54, 15.54, 4.15, 3.9, 3.65, 3.42, 3.2, -1.9, -3.32, -3.97, -3.91, -3.27, 2.38, 2.36, 2.39, 2.46, 2.57, 2.2, 1.93, 1.22, 0.17, -1.06, 4.18, 4.59, 5.04, 5.52, 6.02, -1.49, 0.74, 3.55, 6.75, 10.06, 9.39, 9.95, 10.5, 11.02, 11.52, 16.0, 12.82, 8.47, 3.21, -2.62, 13.61, 13.76, 13.84, 13.84, 13.77, -24.16, -22.03, -17.9, -11.95, -4.55, 11.63, 10.99, 10.27, 9.47, 8.61]])

# change to the size of theta ( 9 ) (for h(x) how many parameters does it have?)
n_params = 9 

best_loss = float('inf')
best_theta = sample_theta(n_params)

for _ in tqdm.tqdm(range(100000)): 
    curr_theta = sample_theta(n_params)
    y_hat = predict(xs, curr_theta)
    curr_loss = get_loss(y_hat, ys)

    if best_loss > curr_loss:
        best_loss = curr_loss
        best_theta = curr_theta
            
        
print("best loss:", best_loss)
print("theta:", best_theta)    
    
    
fig = px.line(x=xs, y=ys, title="f(x) vs Fortuna solution")
fig.add_scatter(x=xs, y=predict(xs, best_theta), mode='lines', name="y_hat")
fig.update_layout(xaxis_range=[xs.min(),xs.max()], yaxis_range=[-6,6])
fig.show()


# to get a solid estimate -> you should train at least 100 models and take the average performance.



100%|██████████| 100000/100000 [00:01<00:00, 83317.52it/s]


best loss: 46.87330433380523
theta: [-3.45460828 -1.03676657 -0.646334    1.57709063 -2.26368578 -1.87801417
  0.51739321  0.74131405 -2.18397316]


CPU times: user 2.24 s, sys: 167 ms, total: 2.41 s
Wall time: 2.52 s


In [5]:
%%time
# BASELINE: measure the regular Fortuna (average best loss over many runs)
# Uses predict, sample_theta, get_loss, xs, ys from the cell above -> run that cell first.

import numpy as np


def fortuna(xs, ys, n_params, n_guesses=100000, predict_fn=predict):
    # Regular Fortuna: n_guesses random guesses, keep the best one
    # predict_fn lets us reuse this function for another model (h(x) later)
    best_loss = float('inf')
    best_theta = None

    for _ in range(n_guesses):
        curr_theta = sample_theta(n_params)
        y_hat = predict_fn(xs, curr_theta)
        curr_loss = get_loss(y_hat, ys)

        if curr_loss < best_loss:
            best_loss = curr_loss
            best_theta = curr_theta

    return best_loss, best_theta


n_runs = 100   

losses = []
for run in range(n_runs):
    loss, theta = fortuna(xs, ys, n_params=9)
    losses.append(loss)
    print(f"run {run+1}/{n_runs}: best loss = {loss:.2f}")

print("-" * 40)
print("average best loss:", np.mean(losses))
print("min / max:", np.min(losses), "/", np.max(losses))

run 1/100: best loss = 48.64
run 2/100: best loss = 49.63
run 3/100: best loss = 39.12
run 4/100: best loss = 46.48
run 5/100: best loss = 36.28
run 6/100: best loss = 50.86
run 7/100: best loss = 18.60
run 8/100: best loss = 59.29
run 9/100: best loss = 56.41
run 10/100: best loss = 12.48
run 11/100: best loss = 40.46
run 12/100: best loss = 50.60
run 13/100: best loss = 31.73
run 14/100: best loss = 35.46
run 15/100: best loss = 34.25
run 16/100: best loss = 44.08
run 17/100: best loss = 54.69
run 18/100: best loss = 31.14
run 19/100: best loss = 67.39
run 20/100: best loss = 40.04
run 21/100: best loss = 22.55
run 22/100: best loss = 58.73
run 23/100: best loss = 45.90
run 24/100: best loss = 45.29
run 25/100: best loss = 34.74
run 26/100: best loss = 28.63
run 27/100: best loss = 50.31
run 28/100: best loss = 43.70
run 29/100: best loss = 50.59
run 30/100: best loss = 53.39
run 31/100: best loss = 38.06
run 32/100: best loss = 30.20
run 33/100: best loss = 28.66
run 34/100: best lo

In [6]:
%%time
# IMPROVED FORTUNA: search around the best theta, with a step size that shrinks over time.
# Big steps at the start (explore), tiny steps at the end (fine-tune).
# We assume that the best theta is somewhere in the neighborhood of the best theta found so far
import numpy as np


def fortuna_shrinking(xs, ys, n_params, n_guesses=100000, step_start=1.0, step_end=0.01, predict_fn=predict):
    # start from one random guess
    best_theta = sample_theta(n_params)
    best_loss = get_loss(predict_fn(xs, best_theta), ys)

    for i in range(n_guesses):
        # step goes linearly from step_start to step_end
        step = step_start + (step_end - step_start) * i / n_guesses

        # draws 9 random numbers centered on 0. Most of them fall between -step and +step
        curr_theta = best_theta + np.random.normal(0, step, size=n_params)
        # keep every parameter inside [-4, 4], same range as sample_theta
        curr_theta = np.clip(curr_theta, -4, 4)
        y_hat = predict_fn(xs, curr_theta)
        curr_loss = get_loss(y_hat, ys)

        if curr_loss < best_loss:
            best_loss = curr_loss
            best_theta = curr_theta

    return best_loss, best_theta


n_runs = 100

losses_shrinking = []
for run in range(n_runs):
    loss, theta = fortuna_shrinking(xs, ys, n_params=9)
    losses_shrinking.append(loss)
    print(f"run {run+1}/{n_runs}: best loss = {loss:.2f}")

print("-" * 40)
print("average best loss:", np.mean(losses_shrinking))
print("min / max:", np.min(losses_shrinking), "/", np.max(losses_shrinking))

# plot the last solution
fig = px.line(x=xs, y=ys, title="f(x) vs improved Fortuna solution")
fig.add_scatter(x=xs, y=predict(xs, theta), mode='lines', name="y_hat")
fig.update_layout(xaxis_range=[xs.min(), xs.max()], yaxis_range=[-6, 6])
fig.show()

run 1/100: best loss = 3.48
run 2/100: best loss = 0.24
run 3/100: best loss = 0.24
run 4/100: best loss = 3.42
run 5/100: best loss = 5.40
run 6/100: best loss = 2.05
run 7/100: best loss = 0.10
run 8/100: best loss = 0.53
run 9/100: best loss = 0.36
run 10/100: best loss = 0.09
run 11/100: best loss = 3.47
run 12/100: best loss = 3.47
run 13/100: best loss = 3.40
run 14/100: best loss = 0.72
run 15/100: best loss = 3.45
run 16/100: best loss = 0.96
run 17/100: best loss = 1.39
run 18/100: best loss = 3.39
run 19/100: best loss = 2.06
run 20/100: best loss = 4.49
run 21/100: best loss = 0.42
run 22/100: best loss = 1.00
run 23/100: best loss = 0.04
run 24/100: best loss = 0.39
run 25/100: best loss = 2.41
run 26/100: best loss = 0.47
run 27/100: best loss = 2.02
run 28/100: best loss = 1.85
run 29/100: best loss = 2.48
run 30/100: best loss = 1.40
run 31/100: best loss = 0.30
run 32/100: best loss = 0.97
run 33/100: best loss = 0.72
run 34/100: best loss = 2.04
run 35/100: best loss =

CPU times: user 2min 16s, sys: 4.75 s, total: 2min 21s
Wall time: 2min 15s


### Part 2 -- What was changed in Fortuna, why, and how it was measured

**What:** instead of drawing every θ uniformly in $[-4, 4]^9$, the improved version is a *local search* around the best θ found so far:
1. Start from one random θ (drawn with `sample_theta`, unchanged).
2. New candidate = `best_theta + N(0, step)`: small Gaussian noise added to every parameter.
3. The candidate is clipped to $[-4, 4]$ so every parameter stays in the allowed range.
4. Keep the candidate only if its loss is lower.
5. `step` shrinks linearly from `1.0` to `0.01` during the 100 000 guesses.

**Why:**
* Regular Fortuna never uses what it has already learned: every guess is independent. In 9 dimensions the space is huge, so hitting a good θ by pure luck is very unlikely.
* The loss is a *continuous* function of θ: if θ is good, a θ close to it is probably good too. Searching near the best θ exploits that.
* Large steps at the start let the search jump between regions (exploration); small steps at the end fine-tune the solution (exploitation).
* Clipping: without it, the search can drift outside $[-4, 4]$. In our tests the unclipped version sometimes got lost (one run ended with loss 123), clipping removed those outliers.

**How it was measured:** both algorithms get the same budget (100 000 guesses) and are run **100 times**; we compare the average, min and max of the best loss (cells above).

| Algorithm (100 runs, 100 000 guesses) | average best loss | min | max |
|---|---|---|---|
| Regular Fortuna | 40.63 | 14.06 | 64.28 |
| Improved, no clipping | 4.57 | 0.02 | 123.64 |
| **Improved, with clipping** | **1.94** | 0.03 | 6.77 |

The improved version is about **20x better** on average and below the "excellent" target of 5.

In [7]:
%%time
# PART 3: optimize h(x) = mu - (zeta * sin(kappa * x)) * (tau * (x + lambda))
# Uses sample_theta, get_loss, fortuna, fortuna_shrinking, xs, ys_h_1, ys_h_2 from the cells above
# -> run cells 17, 18 (only the def is needed) and 19 first.
# Do not run Task 5 before this cell: it redefines predict and xs.
import numpy as np
import plotly.express as px


def predict_h(x, theta):
    # 5 parameters: mu (vertical offset), zeta and tau (amplitude), kappa (frequency), lam (shift)
    mu, zeta, kappa, tau, lam = theta
    return mu - (zeta * np.sin(kappa * x)) * (tau * (x + lam))


def summary(losses):
    # the median is shown too: a few runs stuck in a local minimum can blow up the average
    return (f"average {np.mean(losses):.2f} | median {np.median(losses):.2f} "
            f"| min {np.min(losses):.2f} | max {np.max(losses):.2f}")


n_params_h = 5
n_runs = 100

# ys_h_2 is written with double brackets -> shape (1, 65), flatten it to (65,)
datasets = {"ys_h_1": ys_h_1.flatten(), "ys_h_2": ys_h_2.flatten()}

for name, ys_target in datasets.items():
    losses_base, losses_impr = [], []
    best_base, best_impr = (float('inf'), None), (float('inf'), None)

    for run in range(n_runs):
        loss_b, theta_b = fortuna(xs, ys_target, n_params_h, predict_fn=predict_h)
        loss_i, theta_i = fortuna_shrinking(xs, ys_target, n_params_h, predict_fn=predict_h)
        losses_base.append(loss_b)
        losses_impr.append(loss_i)
        best_base = min(best_base, (loss_b, theta_b), key=lambda t: t[0])
        best_impr = min(best_impr, (loss_i, theta_i), key=lambda t: t[0])

    print("=" * 50)
    print(name)
    print("  regular Fortuna :", summary(losses_base))
    print("  improved Fortuna:", summary(losses_impr))
    print("  best theta (improved) [mu, zeta, kappa, tau, lam]:", np.round(best_impr[1], 3))

    fig = px.line(x=xs, y=ys_target, title=f"{name} vs Fortuna solutions (best of {n_runs} runs)")
    fig.add_scatter(x=xs, y=predict_h(xs, best_base[1]), mode='lines', name="regular Fortuna")
    fig.add_scatter(x=xs, y=predict_h(xs, best_impr[1]), mode='lines', name="improved Fortuna")
    fig.show()

ys_h_1
  regular Fortuna : average 357.40 | median 338.48 | min 38.56 | max 811.60
  improved Fortuna: average 1045.13 | median 0.10 | min 0.01 | max 4354.53
  best theta (improved) [mu, zeta, kappa, tau, lam]: [ 3.898  2.301 -3.7    3.964 -3.5  ]


ys_h_2
  regular Fortuna : average 3014.83 | median 3002.05 | min 2851.43 | max 3312.13
  improved Fortuna: average 3173.08 | median 2833.63 | min 2833.59 | max 3607.90
  best theta (improved) [mu, zeta, kappa, tau, lam]: [ 4.    -1.627 -2.627 -2.767 -3.956]


CPU times: user 5min 16s, sys: 966 ms, total: 5min 17s
Wall time: 5min 16s


### Part 3 -- Does the improved Fortuna also outperform on h(x)? Why?

Model: $h(x) = \mu - (\zeta \sin(\kappa x)) (\tau (x + \lambda))$, so $\theta = \{\mu, \zeta, \kappa, \tau, \lambda\}$ and $|\theta| = 5$.

Results (100 runs each, 100 000 guesses per run):

| Data | Algorithm | average | median | min | max |
|---|---|---|---|---|---|
| `ys_h_1` | Regular Fortuna | 343.66 | 328.27 | 25.49 | 818.36 |
| `ys_h_1` | Improved Fortuna | 827.42 | **0.12** | 0.02 | 4354.49 |
| `ys_h_2` | Regular Fortuna | 3027.50 | 3022.98 | 2870.48 | 3259.02 |
| `ys_h_2` | Improved Fortuna | 3188.47 | **2833.66** | 2833.59 | 3606.18 |

**Answer: yes for a typical run, but not on average.**

* **`ys_h_1`:** in more than half of the runs the improved version finds an almost perfect fit (median 0.12, against 328 for regular Fortuna). But its *average* is worse, because a few runs end with a huge loss (up to 4354).
  * **Why:** local search only accepts improvements, so it follows the loss "downhill" from its starting point. $h(x)$ has many local minima: $\sin(\kappa x)$ repeats for many values of $\kappa$, and the product $\sin \times$ line gives an irregular loss landscape. If the random start is in the wrong "valley", the search never leaves it, especially once the step becomes small.
  * Regular Fortuna does not get stuck (every guess is independent), but it is never precise either: it never gets below 25.
  * Note: $\zeta$ and $\tau$ are redundant, only the product $\zeta\tau$ matters. So there are infinitely many equally good θ.
* **`ys_h_2`:** the data jumps abruptly every 5 points (e.g. 5.48 → 16.58, 15.54 → 4.15). $h(x)$ is smooth and cannot reproduce these jumps, so **no θ gets a low loss**: the best possible is around 2833. The improved version reaches that floor in most runs (median 2833.66, a result regular Fortuna never reaches), but it is again worse on average because of the runs stuck in local minima. The gap between the two algorithms is much smaller here, because even the best θ is a poor fit.

**Conclusion:** local search assumes that "close to a good θ = good θ", with one valley to go down into. That holds for the sum of sines, but $h(x)$ has several valleys, so the improved Fortuna becomes a gamble: very good or very bad. A natural fix would be **restarts** (e.g. 10 local searches of 10 000 guesses each from different random starting points, keeping the best one), which combines the exploration of regular Fortuna with the precision of local search.

# Task 5 -- Fortuna For Decision Trees

Implement a Decision Tree for solving the XOR problem.   
Here: there are 2 real valued inputs, x0, x1  (found in xs).  

And the DT should take these two as input and predict an output: 0 or 1.   
The true answer can be found in ys_true.


In [ ]:
import numpy as np
rng = np.random.default_rng(42)
n_examples = 40

# 40 points ; chaque point = 2 nombres réels dans [0,1] : x0 (colonne 0) et x1 (colonne 1).
xs = rng.uniform(size=(n_examples, 2))

# make a true y
# "grand" = > 0.5. ys_true = XOR : 1 si EXACTEMENT un des deux est grand, sinon 0.
b = (xs>0.5).astype(int)
ys_true = np.logical_xor(b[:, 0], b[:, 1]).astype(int)


##################################
# here you implement a Decision Tree that is built using the fortuna algorithm.
# The depth of the tree should be less than 4
# The DT should reach 100% accuracy.

# -- CODE --

# BUT : trouver un arbre de décision (profondeur 2) qui reproduit XOR sur les 40 points.
# Fortuna = on tire des arbres au hasard et on garde le meilleur (meilleur score).
frt = np.random.default_rng(0)   # générateur aléatoire pour Fortuna (séparé des données)

# Un arbre est fait de nœuds-questions et de feuilles-réponses :
#   - nœud    (= une liste)  : [variable, seuil, branche_vrai, branche_faux]
#       variable : 0 -> x0, 1 -> x1 ;  test : x[variable] >= seuil
#   - feuille (= un nombre)  : 0 ou 1, la réponse finale.


def grow(depth):
    # Construit UN arbre au hasard. Seules les questions sont tirées au hasard ;
    # les feuilles restent vides (None) et seront remplies plus tard par fit().

    # Cas de base : on a atteint la profondeur maximale -> on crée une feuille vide.
    if depth == 0:
        return None

    # Sinon, on crée un nœud-question avec des valeurs tirées au hasard.
    variable = frt.integers(0, 2)      # variable à tester : 0 ou 1 (x0 ou x1)
    seuil = frt.uniform(0, 1)          # valeur de comparaison (entre 0 et 1)

    # On construit les deux sous-arbres, un cran plus bas.
    branche_vrai = grow(depth - 1)     # utilisé si x[variable] >= seuil
    branche_faux = grow(depth - 1)     # utilisé sinon

    return [variable, seuil, branche_vrai, branche_faux]


def fit(arbre, indices):
    # Remplit les feuilles vides par un VOTE MAJORITAIRE :
    # dans chaque feuille, on écrit la vraie réponse (0 ou 1) la plus fréquente
    # parmi les points de "indices" qui arrivent dans cette feuille.

    # Cas de base : on est sur une feuille vide -> on lui donne sa réponse.
    if arbre is None:
        if len(indices) == 0:
            return 0                                     # aucun point -> 0 par défaut
        reponses = ys_true[indices]                      # vraies réponses des points reçus
        reponse_majoritaire = int(reponses.mean() >= 0.5)  # 1 si plus de 1 que de 0
        return reponse_majoritaire

    # Sinon, on est sur un nœud-question -> on répartit les points dans les 2 branches.
    variable, seuil, branche_vrai, branche_faux = arbre
    part_dans_vrai = xs[indices, variable] >= seuil      # masque : qui va dans "vrai"
    indices_vrai = indices[part_dans_vrai]                # points qui vont dans "vrai"
    indices_faux = indices[~part_dans_vrai]               # points qui vont dans "faux"

    # On remplit chaque sous-arbre avec seulement ses points.
    nouveau_vrai = fit(branche_vrai, indices_vrai)
    nouveau_faux = fit(branche_faux, indices_faux)

    return [variable, seuil, nouveau_vrai, nouveau_faux]


def predict(arbre, X):
    # Fait descendre chaque point dans l'arbre jusqu'à une feuille,
    # puis renvoie la réponse lue dans cette feuille (0 ou 1).
    predictions = []

    for point in X:
        noeud = arbre                                    # on démarre à la racine

        # On descend tant qu'on est sur une question (représentée par une liste).
        while isinstance(noeud, list):
            variable, seuil, branche_vrai, branche_faux = noeud
            valeur = point[variable]

            if valeur >= seuil:
                noeud = branche_vrai                     # on va dans la branche "vrai"
            else:
                noeud = branche_faux                     # on va dans la branche "faux"

        # Ici, noeud est une feuille : un simple nombre (0 ou 1).
        predictions.append(noeud)

    return np.array(predictions)


# --- Fortuna : essayer des arbres au hasard et garder le meilleur ---
meilleur_arbre = None
meilleur_score = -1                                      # -1 : pire score possible au départ

for tentative in range(100000):
    arbre_aleatoire = grow(2)                            # 1) tirer un arbre au hasard
    arbre_complet = fit(arbre_aleatoire, np.arange(len(xs)))  # 2) remplir ses feuilles
    predictions = predict(arbre_complet, xs)             # 3) prédire les 40 points
    score = (predictions == ys_true).mean()              # 4) score = bonnes réponses / 40

    if score > meilleur_score:                           # 5) meilleur que le record ?
        meilleur_arbre = arbre_complet
        meilleur_score = score
        if score == 1.0:                                 # 6) parfait -> on s'arrête
            break

print("accuracy:", meilleur_score, "| profondeur: 2 (< 4)")
print("tree:", meilleur_arbre)


accuracy: 1.0 | profondeur: 2 (< 4)
tree: [np.int64(1), 0.4997778837542417, [np.int64(0), 0.5083801066108724, 0, 1], [np.int64(0), 0.4746443596557963, 1, 0]]


# Task 6 -- Best Route App

A company called Knut Knut Transport AS is using one of the 4 routes for delivery (see Pitch\_knutknut\_transport.pdf for more info).
* A -> C -> D  
* A -> C -> E  
* B -> C -> D  
* B -> C -> E  

They have discovered that they can transport the goods faster by picking the right route given 
only the depature time of the transport.

In the file "traffic.jsonl" on Canvas is the collected data up to the current point in time.
So far, they have just selected a route at random, now they want to implement a simple web-api
that can help the drivers select the best route. 

* Using the data found in traffic.jsonl, create a ML model that given an time (hour:min) can select the fastest route to travel. **Be sure to document in a google doc/word/notebook/... how YOU created your ML model.** The ML model should be trained using the Fortuna algorithm.

* Download the knut\_knut\_app.py found on canvas an implement the function: get\_the\_best\_route\_as\_a\_text\_informatic(dep\_hour, dep\_min) such that the web application works and gives a good estimated time.

* The CEO (Knut) want you to estimate how much time they can save from using this new app.
* Prepare a VERY short presentation (pdf) on the gains and tech behind the app (random subsample will present for the class)


Some hints:
* Look closely at the data using both numerical analysis and visual (plotting). 
* Use the python library datetime to handle time calculations. 
* Make sure the features are scaled in a way that make sense.
* Once you have a model working, use the python library pickle to save/load the model.